tollway/events.py

```diff
@@ -1,61 +1,97 @@
-    random_integer = random.randint(TIME_UNIT[time_unit]["min"], TIME_UNIT[time_unit]["max"])
-    current_timestamp = datetime.strptime(ts, TIMESTAMP_FORMAT)
-    return (current_timestamp - timedelta(**{time_unit: random_integer})).strftime(TIMESTAMP_FORMAT)
-
-
-def create_late_event(events_log: EventsLog, fake: Faker, tollways: dict, time_unit: str) -> dict:
-    past_timestamps = events_log["late_events"][time_unit]
-    late_event_tollway = create_tollway(tollways=tollways)
-    late_event_vehicle = create_vehicle(fake)
-    late_event_message = create_message(late_event_vehicle, late_event_tollway)
-    random_ts = random.choice(past_timestamps[slice(0, len(past_timestamps))])
-    late_event_message["timestamp"] = _calculate_late_timestamp(time_unit=time_unit, ts=random_ts)
-    return late_event_message
-
-
-def process_late_event(
-    events_log: EventsLog,
-    fake: Faker,
-    tollways: dict,
-    publisher: PublisherClient,
-    topic_path: Topic,
-    time_unit: str,
-) -> EventsLog:
-    late_event_message = create_late_event(events_log, fake, tollways, time_unit)
-    if publisher and topic_path:
-        data = encode_message(message=late_event_message)
-        future = publisher.publish(topic=topic_path, messages=data)
-    late_event_message["is_late"] = time_unit
-    events_log["all_events"].append(late_event_message)
-    events_log["late_events"][time_unit] = []
-    return events_log
-
-
-def create_duplicate_event(events_log: EventsLog) -> dict[str, str]:
-    duplicate_event = random.choice(events_log["past_events"])
-    return duplicate_event
-
-
-def process_duplicate_event(
-    events_log: EventsLog, publisher: PublisherClient, topic_path: Topic
-) -> EventsLog:
-    duplicate_event = create_duplicate_event(events_log)
-    if publisher and topic_path:
-        data = encode_message(message=duplicate_event)
-        future = publisher.publish(topic=topic_path, messages=data)
-    duplicate_event["is_duplicate"] = "true"
-    events_log["all_events"].append(duplicate_event)
-    events_log["past_events"] = []
-    return events_log
+class EventProcessor:
+    def __init__(self, events_log: EventsLog, publisher: PublisherClient, topic_path: Topic) -> None:
+        self.events_log = events_log
+        self.publisher = publisher
+        self.topic_path = topic_path
+        self.processed = False
+
+    def publish_event(self, event_message: dict[str, str]) -> None:
+        if self.publisher and self.topic_path:
+            data = encode_message(message=event_message)
+            future = self.publisher.publish(topic=self.topic_path, data=data)
+
+    def logging(self, event_message: dict[str, Union[str, bool]]) -> None:
+        self.events_log["all_events"].append(event_message)
+
+    def create_event(self) -> dict[str, Union[str, bool]]:
+        raise NotImplementedError
+
+    def process_event(self) -> tuple[EventsLog, bool]:
+        raise NotImplementedError
```

Comment:
use ABC?

Reply:
https://pybit.es/articles/elevate-your-python-harnessing-the-power-of-abstract-base-classes-abcs/

Reply:
now superseded (and hot off the press):
https://pybit.es/articles/typing-protocol-abc-alternative/

Reply:
aware of but never used ABC before. Will create issue and work on it.